test/jdk/java/nio/file/Files/CopyInterference.java

```diff
@@ -0,0 +1,153 @@
+            this.source = source;
+            this.target = target;
+            this.options = options;
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (running.get()) {
+                    Files.copy(source, target, options);
+                }
+            } catch (FileAlreadyExistsException e) {
+                running.set(false);
+                throw new RuntimeException("Unexpected exception", e);
+            } catch (FileSystemException e) {
+                System.out.printf("Expected FileSystemException: \"%s\"%n",
+                                  e.getMessage());
+            } catch (IOException e) {
+                running.set(false);
+                throw new RuntimeException("Unexpected exception", e);
+            }
+            running.set(false);
+        }
+    }
+
+    private static Stream<Arguments> pathAndOptionsProvider()
+        throws IOException {
+        Path parent = Path.of(System.getProperty("test.dir", "."));
+        Path dir = Files.createTempDirectory(parent, "foobargus");
+        dir.toFile().deleteOnExit();
+
+        List<Arguments> list = new ArrayList<Arguments>();
+
+        // regular file
+        Path sourceFile = Files.createTempFile(dir, "foo", "baz");
+        Class c = CopyInterference.class;
+        String name = "CopyInterference.class";
+
+        try (InputStream in = c.getResourceAsStream(name)) {
+            Files.copy(in, sourceFile, REPLACE_EXISTING);
+        }
+
+        Arguments args = Arguments.of(sourceFile, dir.resolve("targetFile"),
+                                      new CopyOption[] {REPLACE_EXISTING});
+        list.add(args);
+
+        // directory
+        Path sourceDirectory = Files.createTempDirectory(dir, "fubar");
+        args = Arguments.of(sourceDirectory, dir.resolve("targetDir"),
+                            new CopyOption[] {REPLACE_EXISTING});
+        list.add(args);
+
+        // symblic link, followed
+        Path link = dir.resolve("link");
+        Files.createSymbolicLink(link, sourceFile);
+        args = Arguments.of(link, dir.resolve("linkFollowed"),
+                            new CopyOption[] {REPLACE_EXISTING});
+        list.add(args);
+
+        // symblic link, not followed
+        args = Arguments.of(link, dir.resolve("linkNotFollowed"),
+                            new CopyOption[] {REPLACE_EXISTING, NOFOLLOW_LINKS});
+        list.add(args);
+
+        return list.stream();
+    }
+
+    @ParameterizedTest
+    @MethodSource("pathAndOptionsProvider")
+    void copy(Path source, Path target, CopyOption[] options)
+        throws InterruptedException, IOException {
+
+        ExecutorService es = Executors.newFixedThreadPool(N_THREADS);
+        CopyTask copyTask = new CopyTask(source, target, options);
+        Future<?>[] results = new Future<?>[N_THREADS];
+        for (int i = 0; i < N_THREADS; i++)
+            results[i] = es.submit(copyTask);
+
+        es.close();
```

Comment:
You can use try-with-resources with ExecutorService to ensure that it is closed the event of an exception.

Reply:
So changed in 541f36b2aa1c2f530cf1280e9b68628d6ff1c869.